### PR TITLE
[Extension] Implement ECS Observer

### PR DESCRIPTION
**Pull request by theRoughCode**
## Description
This PR implements the ECS Observer that queries the ECS/EC2 API to discover scrape targets within a configured ECS cluster.

There are 2 modes to discover the targets based on the config:
* **Mode 1: Docker label-Based:** Add docker labels to the containers to indicate the port and metric path. The ECS Observer can then be configured to discover targets with the matching docker label and container port.
* **Mode 2: ECS Task Definition ARN-based:**  The ECS Observer can be configured to discover targets when the target's ECS task definition ARN matches the configured regex and it matches the configured container ports.

Two modes can be enabled together and the ECS Observer will de-duplicate the discovered targets based on: *{private_ip}:{port}/{metrics_path}*

## Implementation Overview

1. The ECS Observer's discovery is triggered by `EndpointsListers:ListEndpoints`.
2. The `TaskRetrievalProcessor` will query the `ECS:ListTasks` API to retrieve the list of current running ECS task ARNs for the specified ECS cluster, and each task will be retrieved using `ECS:DescribeTasks`.
3. The `TaskDefinitionProcessor` will get the ECS Task Definition from LRU cache or, if there is none in cache, call `ECS:DescribeTaskDefinition` and cache. LRU cache size (2000) based on [ECS service quota](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-quotas.html)
4. The `TaskFilterProcessor` will filter out tasks based on whether they match the configured docker label-based SD config or task definition ARN-based SD config.
5. The `MetadataProcessor` will get the containerInstance/ec2 instance info from LRU cache if the tasks is running on EC2 launch type. LRU cache size (2000) based on [ECS service quota](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-quotas.html). If not cached, call `ECS: DescribeContainerInstances` / `EC2:DescribeInstances` with batch size = 100.
6. If `sd_result_file` is configured, the list of discovered targets will be written to the configured file path. Otherwise, `EndpointsListers:ListEndpoints` will return the list of discovered endpoints and all subscribers will be notified of the updated endpoints.

**Note:** The main reason for enabling writing to a static file is because the current implementation of the observer/receiver creator framework is less efficient than using a single receiver to scrape from multiple targets. The related discussion can be found in #1395.

**Link to tracking Issue:** #1395 

**Testing:**
Unit tests were added and ran successfully. A test setup in AWS ECS was also created to verify the implementation.
*Note:* Functions involving the ECS/EC2 API don't have unit tests. A next step would be to add mocked APIs to test these functions.

**Documentation:**
A new README doc was added for the `ecs_observer`, and the `observer`/`receiver_creator` READMEs were modified.

### CODE REVIEW

**Comment by theRoughCode** [comment]
A comprehensive testing suite has not been added yet. I created this PR to get feedback on the approach. Making this a draft PR until I add in the relevant unit tests.

**Comment by theRoughCode** [comment]
cc: @mxiamxia for feedback

**Comment by tigrannajaryan** [comment]
@jrcamp please review this.

**Review by jrcamp — CHANGES_REQUESTED** [review]
@theRoughCode sorry for delayed review.

Can you follow https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#how-to-structure-prs-to-get-expedient-reviews to break this up into more manageable pieces? Overall it's on a good track though.

I'd suggest doing a PR with just README.md changes first so we can iterate there first if needed to avoid unnecessary code churn.

**Comment by bogdandrutu** [comment]
@theRoughCode friendly ping :)

**Comment by pingleig** [comment]
Hi @bogdandrutu @jrcamp sorry for not responding the comments. I am from @theRoughCode team at AWS, he has finished his internship and I got this task from @mxiamxia (prometheus service discovery using ECS API).

Just want to have some clarification before I start working on it

- The overall implementation in this PR works? I saw there were discussion https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1395#issuecomment-747715986 and we decided not to use the observer/receiver creator framework for performance issues and use the current approach.
- split the pr two three prs? one for design (in README.md), one for impl, one for enabling the impl after test?

**Comment by theRoughCode** [comment]
@bogdandrutu @jrcamp apologies for not responding earlier, I've been super busy with work the past month and was hoping to get to this once stuff settled. But it looks like @pingleig is taking over, so I'll hand it over to him!

@pingleig that's right, from our discussion a month back, we decided not to use the observer/creator framework _for now_ until its performance is up to par with the "one receiver" method. 

I've added some documentation in this PR already that should help in drafting up the README for the initial PR that @jrcamp proposed. Feel free to ping me if you have any questions!

**Comment by tigrannajaryan** [comment]
@pingleig do you plan to continue working on this?

File: extension/observer/ecsobserver/README.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+| port_label          | Mandatory   | container's docker label name that specifies the metrics port                   |
+| job_name_label      | Optional    | container's docker label name that specifies the scrape job name. (Default: "") |
+| metrics_path_label  | Optional    | container's docker label name that specifies the metrics path. (Default: "")    |
+
+#### Task Definition ARN-Based Service Discovery Configuration
+
+| Name                |             | Description                                                   |
+|---------------------|-------------|---------------------------------------------------------------|
+| task_definition_arn_pattern    | Mandatory   | Regex pattern to match against ECS task definition ARN |
+| metrics_ports                  | Mandatory   | container ports separated by semicolon. Only containers that expose these ports will be discovered   |
```
Comment (jrcamp) [inline extension/observer/ecsobserver/README.md:37]: Why not a list?

**Review by jrcamp — COMMENTED** [review]
@pingleig README looks good overall just a couple notes. Can you start breaking it up based on https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#how-to-structure-prs-to-get-expedient-reviews?

**Comment by pingleig** [comment]
Hi @jrcamp I just created a new PR that only contains the doc https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2377 I will address the comment you have in current PR in the new one.

**Comment by pingleig** [comment]
@theRoughCode I think we can close this one as the content is being ported into other (smaller) PRs

**Comment by theRoughCode** [comment]
@pingleig sounds good! Thanks ☺️

**Pull request closed without merging**